### PR TITLE
fix(codewhisperer): latency caused by credentials

### PR DESCRIPTION
**Pull request by zixlin7**
## Problem
fix the latency introduced by not cacheing the credentials
## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by leigaol — APPROVED** [review]

**Comment by leigaol** [comment]
Please fix the error in unit test

**Comment by zixlin7** [comment]
> Please fix the error in unit test

I don't think it's related, is this known flaky?

**Review by andrewyuq — APPROVED** [review]

**Comment by leigaol** [comment]
>I don't think it's related, is this known flaky? We should probably track if not

Let me fix it in a different PR

File: src/codewhisperer/util/cognitoIdentity.ts [hunk 1]
```diff
@@ -29,7 +29,6 @@ export const getCognitoCredentials = async (): Promise<CognitoIdentityCredential
         }
 
         const credentials = new CognitoIdentityCredentials({ IdentityId: identityId }, { region })
-        await credentials.getPromise()
```
Comment (leigaol) [inline src/codewhisperer/util/cognitoIdentity.ts:32]: Can you justify removing this line? Thanks!

**Comment by leigaol** [comment]
You can add a bug fix change log item.

File: src/codewhisperer/client/codewhisperer.ts [hunk 2]
```diff
@@ -39,8 +39,16 @@ export type ListCodeScanFindingsRequest = Readonly<CodeWhispererClient.ListCodeS
...
 export type ArtifactMap = Readonly<CodeWhispererClient.ArtifactMap>
 export class DefaultCodeWhispererClient {
+    private credentials?: CognitoIdentityCredentials
+
     private async createSdkClient(): Promise<CodeWhispererClient> {
-        const credentials = !isCloud9() ? await getCognitoCredentials() : undefined
+        if (!this.credentials && !isCloud9()) {
+            this.credentials = await getCognitoCredentials()
+        } else if (this.credentials?.needsRefresh()) {
+            this.credentials.refresh(error => {
```
Comment (leigaol) [inline src/codewhisperer/client/codewhisperer.ts:48]: 2 changes are needed:
1. Catch all exceptions when either getting the credentials or refreshing the credentials. Throw an Error and a followup PR is needed to clean up telemetry logic when cognito API went wrong.
2.  Wait until this.credentials.refresh is finished.

**Review by leigaol — APPROVED** [review]

**Pull request closed without merging**